File: room_exam/app/src/main/java/com/hyeonwoo/room_exam/MainViewModel.java
package com.hyeonwoo.room_exam;

import android.app.Application;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.room.Room;

import java.util.List;

public class MainViewModel extends AndroidViewModel {

    AppDatabase db;

    public MainViewModel(@NonNull Application application) {
        super(application);

        db = Room.databaseBuilder(application, AppDatabase.class, "todo-db")
//                .allowMainThreadQueries() // MainThread에서 db조작사용 (실무에서는 사용하지 않는다) -> asyncTask로 대체
                .build(); // db객체를 생성
    }

    public LiveData<List<Todo>> getAll() {
        return db.todoDao().getAll();
    }

    public void insert(Todo todo) {
        new InsertAsyncTask(db.todoDao()).execute(todo);
    }

    private static class InsertAsyncTask extends AsyncTask<Todo, Void, Void> {
        private TodoDao mTodoDao;

        public InsertAsyncTask (TodoDao todoDao) {
            this.mTodoDao = todoDao;
        }

        @Override
        protected Void doInBackground(Todo... todos) {
            mTodoDao.insert(todos[0]);
            return null;
        }
    }
}

File: room_exam_kotlin/app/src/main/java/com/hyeonwoo/room_exam_kotlin/MainViewModel.kt
package com.hyeonwoo.room_exam_kotlin

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.room.Room

class MainViewModel(application: Application): AndroidViewModel(application) {
    private val db = Room.databaseBuilder(application, AppDatabase::class.java, "tood-db").build()

    var todos : LiveData<List<Todo>>

    init {
        todos = getAll()
    }

    fun getAll(): LiveData<List<Todo>> {
        return db.todoDao().getAll();
    }
    /*
     suspend 한정자는 해당 함수가 무조건 코루틴 안에서 처리되어야 한다는 뜻으로,
     코루틴 스코프를 벗어나면 컴파일러에서 에러로 처리한다.
     */
    suspend fun insert(todo : Todo) {
        db.todoDao().insert(todo)
    }


}